// node-kakao/dist/talk/struct/api/open/open-post-struct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenPostApiStruct = exports.OpenPostReactNotiStruct = exports.OpenPostReactStruct = exports.OpenPostListStruct = exports.OpenPostStruct = exports.OpenPostReactionStruct = exports.OpenPostDescStruct = exports.OpenPostTagStruct = exports.OpenPostTagType = void 0;
const json_proxy_mapper_1 = require("json-proxy-mapper");
const json_util_1 = require("../../../../util/json-util");
var OpenPostTagType;
(function (OpenPostTagType) {
    OpenPostTagType[OpenPostTagType["SHARP"] = 2] = "SHARP";
})(OpenPostTagType = exports.OpenPostTagType || (exports.OpenPostTagType = {}));
var OpenPostTagStruct;
(function (OpenPostTagStruct) {
    OpenPostTagStruct.Mappings = {
        type: 't',
        text: 'c'
    };
    OpenPostTagStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenPostTagStruct.Mappings);
})(OpenPostTagStruct = exports.OpenPostTagStruct || (exports.OpenPostTagStruct = {}));
var OpenPostDescStruct;
(function (OpenPostDescStruct) {
    OpenPostDescStruct.Mappings = {
        text: 'text',
        tagList: 'tags'
    };
    OpenPostDescStruct.ConvertMap = {
        tagList: new json_proxy_mapper_1.Converter.Array(OpenPostTagStruct.Mappings)
    };
    OpenPostDescStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenPostDescStruct.Mappings);
})(OpenPostDescStruct = exports.OpenPostDescStruct || (exports.OpenPostDescStruct = {}));
var OpenPostReactionStruct;
(function (OpenPostReactionStruct) {
    OpenPostReactionStruct.Mappings = {
        reactionId: 'reactId',
        linkId: 'linkId',
        type: 'type',
        name: 'name',
        description: 'description',
        profileImageURL: 'profileImagePath'
    };
    OpenPostReactionStruct.ConvertMap = {
        linkId: json_util_1.JsonUtil.LongConverter,
        reactionId: json_util_1.JsonUtil.LongConverter
    };
    OpenPostReactionStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenPostReactionStruct.Mappings);
})(OpenPostReactionStruct = exports.OpenPostReactionStruct || (exports.OpenPostReactionStruct = {}));
var OpenPostStruct;
(function (OpenPostStruct) {
    OpenPostStruct.Mappings = {
        id: 'id',
        linkId: 'linkId',
        description: 'postDescription',
        postDataList: 'postDatas',
        date: 'date',
        postURL: 'postUrl',
        reactionList: 'reacts',
        reactionUserList: 'reactUsers',
        scrapData: 'scrapData',
        latestUpdateToken: 'latestUpdateToken'
    };
    OpenPostStruct.ConvertMap = {
        linkId: json_util_1.JsonUtil.LongConverter,
        description: new json_proxy_mapper_1.Converter.Object(OpenPostDescStruct.Mappings, OpenPostDescStruct.ConvertMap),
        reactionUserList: new json_proxy_mapper_1.Converter.Array(OpenPostReactionStruct.Mappings, OpenPostReactionStruct.ConvertMap),
    };
    OpenPostStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenPostStruct.Mappings, OpenPostStruct.ConvertMap);
})(OpenPostStruct = exports.OpenPostStruct || (exports.OpenPostStruct = {}));
var OpenPostListStruct;
(function (OpenPostListStruct) {
    OpenPostListStruct.Mappings = {
        count: 'count',
        postList: 'posts'
    };
    OpenPostListStruct.ConvertMap = {
        postList: new json_proxy_mapper_1.Converter.Array(OpenPostStruct.Mappings, OpenPostStruct.ConvertMap)
    };
    OpenPostListStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenPostListStruct.Mappings, OpenPostListStruct.ConvertMap);
})(OpenPostListStruct = exports.OpenPostListStruct || (exports.OpenPostListStruct = {}));
var OpenPostReactStruct;
(function (OpenPostReactStruct) {
    OpenPostReactStruct.Mappings = {
        postId: 'postId',
    };
    OpenPostReactStruct.ConvertMap = {
        postId: json_util_1.JsonUtil.LongConverter
    };
    OpenPostReactStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenPostReactStruct.Mappings);
})(OpenPostReactStruct = exports.OpenPostReactStruct || (exports.OpenPostReactStruct = {}));
var OpenPostReactNotiStruct;
(function (OpenPostReactNotiStruct) {
    OpenPostReactNotiStruct.Mappings = {
        linkIdList: 'linkIds',
    };
    OpenPostReactNotiStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenPostReactNotiStruct.Mappings);
})(OpenPostReactNotiStruct = exports.OpenPostReactNotiStruct || (exports.OpenPostReactNotiStruct = {}));
var OpenPostApiStruct;
(function (OpenPostApiStruct) {
    OpenPostApiStruct.Mappings = {
        post: 'post',
    };
    OpenPostApiStruct.ConvertMap = {
        post: new json_proxy_mapper_1.Converter.Object(OpenPostStruct.Mappings, OpenPostStruct.ConvertMap)
    };
    OpenPostApiStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenPostApiStruct.Mappings, OpenPostApiStruct.ConvertMap);
})(OpenPostApiStruct = exports.OpenPostApiStruct || (exports.OpenPostApiStruct = {}));
