// node-kakao/dist/talk/struct/api/open/open-recommend-struct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenPostRecommendStruct = exports.OpenRecommendStruct = exports.OpenRecommend = void 0;
const json_proxy_mapper_1 = require("json-proxy-mapper");
const open_search_struct_1 = require("./open-search-struct");
var OpenRecommend;
(function (OpenRecommend) {
    let TagItem;
    (function (TagItem) {
        TagItem.Mappings = {
            tagName: 'tn',
            backgroundColor: 'bgColor',
            pageReferrer: 'apr'
        };
        TagItem.MAPPER = new json_proxy_mapper_1.ObjectMapper(TagItem.Mappings);
    })(TagItem = OpenRecommend.TagItem || (OpenRecommend.TagItem = {}));
    let LinkItem;
    (function (LinkItem) {
        LinkItem.Mappings = {
            linkName: 'ln',
            linkType: 'lt',
            linkURL: 'lu',
            linkImageURL: 'liu',
            memberCount: 'mcnt',
            description: 'desc',
            ownerNickname: 'nn',
            ownerProfileImageURL: 'pi',
            lastChatAt: 'writtenAt',
            reactCount: 'rc',
            pageReferrer: 'apr'
        };
        LinkItem.MAPPER = new json_proxy_mapper_1.ObjectMapper(LinkItem.Mappings);
    })(LinkItem = OpenRecommend.LinkItem || (OpenRecommend.LinkItem = {}));
    let TagList;
    (function (TagList) {
        TagList.Mappings = {
            title: 'title',
            items: 'items'
        };
        TagList.ConvertMap = {
            items: new json_proxy_mapper_1.Converter.Array(TagItem.Mappings)
        };
        TagList.MAPPER = new json_proxy_mapper_1.ObjectMapper(TagList.Mappings, TagList.ConvertMap);
    })(TagList = OpenRecommend.TagList || (OpenRecommend.TagList = {}));
    let LinkList;
    (function (LinkList) {
        LinkList.Mappings = {
            title: 'title',
            items: 'items'
        };
        LinkList.ConvertMap = {
            items: new json_proxy_mapper_1.Converter.Array(LinkItem.Mappings)
        };
        LinkList.MAPPER = new json_proxy_mapper_1.ObjectMapper(LinkList.Mappings, LinkList.ConvertMap);
    })(LinkList = OpenRecommend.LinkList || (OpenRecommend.LinkList = {}));
})(OpenRecommend = exports.OpenRecommend || (exports.OpenRecommend = {}));
var OpenRecommendStruct;
(function (OpenRecommendStruct) {
    OpenRecommendStruct.Mappings = {
        tag: 'tag',
        linkList: 'links'
    };
    OpenRecommendStruct.ConvertMap = {
        tag: new json_proxy_mapper_1.Converter.Object(OpenRecommend.TagList.Mappings, OpenRecommend.TagList.ConvertMap),
        linkList: new json_proxy_mapper_1.Converter.Array(OpenRecommend.LinkList.Mappings, OpenRecommend.LinkList.ConvertMap),
    };
    OpenRecommendStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenRecommendStruct.Mappings, OpenRecommendStruct.ConvertMap);
})(OpenRecommendStruct = exports.OpenRecommendStruct || (exports.OpenRecommendStruct = {}));
var OpenPostRecommendStruct;
(function (OpenPostRecommendStruct) {
    OpenPostRecommendStruct.Mappings = {
        title: 'title',
        postList: 'recommendPosts'
    };
    OpenPostRecommendStruct.ConvertMap = {
        postList: new json_proxy_mapper_1.Converter.Array(open_search_struct_1.OpenSearchStruct.PostItem.Mappings, open_search_struct_1.OpenSearchStruct.PostItem.ConvertMap),
    };
    OpenPostRecommendStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenPostRecommendStruct.Mappings, OpenPostRecommendStruct.ConvertMap);
})(OpenPostRecommendStruct = exports.OpenPostRecommendStruct || (exports.OpenPostRecommendStruct = {}));
