// node-kakao/dist/talk/struct/api/open/open-search-struct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenPostSearchStruct = exports.OpenSearchStruct = exports.OpenSearchType = void 0;
const json_proxy_mapper_1 = require("json-proxy-mapper");
const open_post_struct_1 = require("./open-post-struct");
const json_util_1 = require("../../../../util/json-util");
var OpenSearchType;
(function (OpenSearchType) {
    OpenSearchType["PROFILE"] = "p";
    OpenSearchType["DIRECT"] = "d";
    OpenSearchType["GROUP"] = "m";
})(OpenSearchType = exports.OpenSearchType || (exports.OpenSearchType = {}));
var OpenSearchStruct;
(function (OpenSearchStruct) {
    let LinkItem;
    (function (LinkItem) {
        LinkItem.Mappings = {
            linkName: 'ln',
            linkType: 'lt',
            linkURL: 'lu',
            linkImageURL: 'liu',
            memberCount: 'mcnt',
            description: 'desc',
            tagList: 'tags',
            locked: 'lk',
            ownerNickname: 'nn',
            ownerProfileImageURL: 'pi',
            lastChatAt: 'writtenAt',
            reactCount: 'rc',
        };
        LinkItem.MAPPER = new json_proxy_mapper_1.ObjectMapper(LinkItem.Mappings);
    })(LinkItem = OpenSearchStruct.LinkItem || (OpenSearchStruct.LinkItem = {}));
    let PostItem;
    (function (PostItem) {
        PostItem.Mappings = {
            id: 'postId',
            linkId: 'linkId',
            nickname: 'nickname',
            profileImageURL: 'profileImagePath',
            description: 'postDescription',
            postDataList: 'postDatas',
            reactionList: 'reacts'
        };
        PostItem.ConvertMap = {
            linkId: json_util_1.JsonUtil.LongConverter,
            description: new json_proxy_mapper_1.Converter.Object(open_post_struct_1.OpenPostDescStruct.Mappings, open_post_struct_1.OpenPostDescStruct.ConvertMap)
        };
        PostItem.MAPPER = new json_proxy_mapper_1.ObjectMapper(PostItem.Mappings);
    })(PostItem = OpenSearchStruct.PostItem || (OpenSearchStruct.PostItem = {}));
    OpenSearchStruct.Mappings = {
        page: 'page',
        count: 'count',
        totalCount: 'totalCount',
        itemList: 'items',
        pageReferrer: 'apr'
    };
    OpenSearchStruct.ConvertMap = {
        itemList: new json_proxy_mapper_1.Converter.Array(LinkItem.Mappings)
    };
    OpenSearchStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenSearchStruct.Mappings, OpenSearchStruct.ConvertMap);
})(OpenSearchStruct = exports.OpenSearchStruct || (exports.OpenSearchStruct = {}));
var OpenPostSearchStruct;
(function (OpenPostSearchStruct) {
    OpenPostSearchStruct.Mappings = {
        page: 'page',
        count: 'count',
        totalCount: 'totalCount',
        postList: 'postItems',
        pageReferrer: 'apr'
    };
    OpenPostSearchStruct.ConvertMap = {
        postList: new json_proxy_mapper_1.Converter.Array(OpenSearchStruct.PostItem.Mappings, OpenSearchStruct.PostItem.ConvertMap)
    };
    OpenPostSearchStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenPostSearchStruct.Mappings, OpenPostSearchStruct.ConvertMap);
})(OpenPostSearchStruct = exports.OpenPostSearchStruct || (exports.OpenPostSearchStruct = {}));
