// node-kakao/dist/talk/struct/api/open/open-preset-struct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenPresetStruct = exports.OpenPresetItemStruct = void 0;
const json_proxy_mapper_1 = require("json-proxy-mapper");
var OpenPresetItemStruct;
(function (OpenPresetItemStruct) {
    OpenPresetItemStruct.Mappings = {
        id: 'id',
        linkImagePath: 'lip',
        linkImageURL: 'liu'
    };
    OpenPresetItemStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenPresetItemStruct.Mappings);
})(OpenPresetItemStruct = exports.OpenPresetItemStruct || (exports.OpenPresetItemStruct = {}));
var OpenPresetStruct;
(function (OpenPresetStruct) {
    OpenPresetStruct.Mappings = {
        preset: 'preset'
    };
    OpenPresetStruct.ConvertMap = {
        preset: new json_proxy_mapper_1.Converter.Object(OpenPresetItemStruct.Mappings)
    };
    OpenPresetStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(OpenPresetStruct.Mappings, OpenPresetStruct.ConvertMap);
})(OpenPresetStruct = exports.OpenPresetStruct || (exports.OpenPresetStruct = {}));
