// node-kakao/dist/packet/media/packet-mini.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketMiniRes = exports.PacketMiniReq = void 0;
const bson_1 = require("bson");
const loco_bson_packet_1 = require("../loco-bson-packet");
const media_request_base_packet_1 = require("./media-request-base-packet");
class PacketMiniReq extends media_request_base_packet_1.MediaRequestBasePacket {
    constructor(Key = '', Offset = 0, ChannelId = bson_1.Long.ZERO, Width = 0, Height = 0, userId = bson_1.Long.ZERO, os = '', version = '', networkType = 0, networkMccMnc = '') {
        super(userId, os, version, networkType, networkMccMnc);
        this.Key = Key;
        this.Offset = Offset;
        this.ChannelId = ChannelId;
        this.Width = Width;
        this.Height = Height;
    }
    get PacketName() {
        return 'MINI';
    }
    toBodyJson() {
        let obj = {
            'k': this.Key,
            'c': this.ChannelId,
            'o': this.Offset,
            'w': this.Width,
            'h': this.Height
        };
        return Object.assign(obj, super.toBodyJson());
    }
}
exports.PacketMiniReq = PacketMiniReq;
class PacketMiniRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, Size = 0) {
        super(status);
        this.Size = Size;
    }
    get PacketName() {
        return 'MINI';
    }
    readBodyJson(rawData) {
        this.Size = rawData['s'];
    }
}
exports.PacketMiniRes = PacketMiniRes;
