// node-kakao/dist/packet/loco-packet-base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatusCode = void 0;
var StatusCode;
(function (StatusCode) {
    StatusCode[StatusCode["SUCCESS"] = 0] = "SUCCESS";
    StatusCode[StatusCode["INVALID_USER"] = -1] = "INVALID_USER";
    StatusCode[StatusCode["INVALID_METHOD"] = -202] = "INVALID_METHOD";
    StatusCode[StatusCode["INVALID_PARAMETER"] = -203] = "INVALID_PARAMETER";
    StatusCode[StatusCode["INVALID_CHATROOM_OPERATION"] = -401] = "INVALID_CHATROOM_OPERATION";
    StatusCode[StatusCode["CHAT_BLOCKED_BY_FRIEND"] = -402] = "CHAT_BLOCKED_BY_FRIEND";
    StatusCode[StatusCode["BLOCKED_IP"] = -444] = "BLOCKED_IP";
    StatusCode[StatusCode["OPERATION_DENIED"] = -500] = "OPERATION_DENIED";
    StatusCode[StatusCode["INVALID_ACCESSTOKEN"] = -950] = "INVALID_ACCESSTOKEN";
    StatusCode[StatusCode["BLOCKED_ACCOUNT"] = -997] = "BLOCKED_ACCOUNT";
    StatusCode[StatusCode["AUTH_REQUIRED"] = -998] = "AUTH_REQUIRED";
    StatusCode[StatusCode["UPDATE_REQUIRED"] = -999] = "UPDATE_REQUIRED";
    StatusCode[StatusCode["SERVER_UNDER_MAINTENANCE"] = -9797] = "SERVER_UNDER_MAINTENANCE";
})(StatusCode = exports.StatusCode || (exports.StatusCode = {}));
