// node-kakao/dist/packet/packet-sync-link.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketSyncLinkRes = exports.PacketSyncLinkReq = void 0;
const open_link_struct_1 = require("../talk/struct/open/open-link-struct");
const loco_bson_packet_1 = require("./loco-bson-packet");
const json_proxy_mapper_1 = require("json-proxy-mapper");
class PacketSyncLinkReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(OpenChatToken = 0) {
        super();
        this.OpenChatToken = OpenChatToken;
    }
    get PacketName() {
        return 'SYNCLINK';
    }
    toBodyJson() {
        let obj = {
            'ltk': this.OpenChatToken
        };
        return obj;
    }
}
exports.PacketSyncLinkReq = PacketSyncLinkReq;
class PacketSyncLinkRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, LinkList = [], IdList = [], OpenChatToken = 0) {
        super(status);
        this.LinkList = LinkList;
        this.IdList = IdList;
        this.OpenChatToken = OpenChatToken;
    }
    get PacketName() {
        return 'SYNCLINK';
    }
    readBodyJson(body) {
        this.LinkList = [];
        if (body['ols']) {
            for (let rawStruct of body['ols']) {
                this.LinkList.push(json_proxy_mapper_1.Serializer.deserialize(rawStruct, open_link_struct_1.OpenLinkStruct.MAPPER));
            }
        }
        if (body['dlis'])
            this.IdList = body['dlis'];
        this.OpenChatToken = body['ltk'];
    }
}
exports.PacketSyncLinkRes = PacketSyncLinkRes;
