// node-kakao/dist/packet/packet-chat-member.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketChatMemberRes = exports.PacketChatMemberReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const member_struct_1 = require("../talk/struct/member-struct");
const json_util_1 = require("../util/json-util");
const bson_1 = require("bson");
const json_proxy_mapper_1 = require("json-proxy-mapper");
const open_link_struct_1 = require("../talk/struct/open/open-link-struct");
class PacketChatMemberReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(ChannelId = bson_1.Long.ZERO, UserIdLIst = []) {
        super();
        this.ChannelId = ChannelId;
        this.UserIdLIst = UserIdLIst;
    }
    get PacketName() {
        return 'MEMBER';
    }
    toBodyJson() {
        return {
            'chatId': this.ChannelId,
            'memberIds': this.UserIdLIst
        };
    }
}
exports.PacketChatMemberReq = PacketChatMemberReq;
class PacketChatMemberRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, ChannelId = bson_1.Long.ZERO, MemberList = []) {
        super(status);
        this.ChannelId = ChannelId;
        this.MemberList = MemberList;
    }
    get PacketName() {
        return 'MEMBER';
    }
    readBodyJson(rawData) {
        this.ChannelId = json_util_1.JsonUtil.readLong(rawData['chatId']);
        this.MemberList = [];
        if (rawData['members']) {
            let memberList = rawData['members'];
            for (let rawMem of rawData['m']) {
                if (rawMem[open_link_struct_1.OpenMemberStruct.Mappings.openToken])
                    this.MemberList.push(json_proxy_mapper_1.Serializer.deserialize(rawMem, open_link_struct_1.OpenMemberStruct.MAPPER));
                else
                    this.MemberList.push(json_proxy_mapper_1.Serializer.deserialize(rawMem, member_struct_1.MemberStruct.MAPPER));
            }
        }
    }
}
exports.PacketChatMemberRes = PacketChatMemberRes;
