// node-kakao/dist/packet/packet-add-member.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketAddMemberRes = exports.PacketAddMemberReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const bson_1 = require("bson");
class PacketAddMemberReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(ChannelId = bson_1.Long.ZERO, MemberIdList = []) {
        super();
        this.ChannelId = ChannelId;
        this.MemberIdList = MemberIdList;
    }
    get PacketName() {
        return 'ADDMEM';
    }
    toBodyJson() {
        return {
            'chatId': this.ChannelId,
            'memberIds': this.MemberIdList
        };
    }
}
exports.PacketAddMemberReq = PacketAddMemberReq;
class PacketAddMemberRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, ChannelId = bson_1.Long.ZERO, MemberIdList = []) {
        super(status);
        this.ChannelId = ChannelId;
        this.MemberIdList = MemberIdList;
    }
    get PacketName() {
        return 'ADDMEM';
    }
    readBodyJson(rawData) {
    }
}
exports.PacketAddMemberRes = PacketAddMemberRes;
