// node-kakao/dist/packet/packet-reaction-count.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketReactionCountRes = exports.PacketReactionCountReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const bson_1 = require("bson");
const open_link_struct_1 = require("../talk/struct/open/open-link-struct");
class PacketReactionCountReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(LinkId = bson_1.Long.ZERO) {
        super();
        this.LinkId = LinkId;
    }
    get PacketName() {
        return 'REACTCNT';
    }
    toBodyJson() {
        return {
            'li': this.LinkId
        };
    }
}
exports.PacketReactionCountReq = PacketReactionCountReq;
class PacketReactionCountRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, LinkId = bson_1.Long.ZERO, ReactType = open_link_struct_1.LinkReactionType.NORMAL, ReactionCount = bson_1.Long.ZERO) {
        super(status);
        this.LinkId = LinkId;
        this.ReactType = ReactType;
        this.ReactionCount = ReactionCount;
    }
    get PacketName() {
        return 'REACTCNT';
    }
    readBodyJson(rawData) {
        this.LinkId = rawData['li'];
        this.ReactType = rawData['rt'];
        this.ReactionCount = rawData['rc'];
    }
}
exports.PacketReactionCountRes = PacketReactionCountRes;
