// node-kakao/dist/packet/packet-login.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketLoginRes = exports.PacketLoginReq = void 0;
const bson_1 = require("bson");
const json_proxy_mapper_1 = require("json-proxy-mapper");
const channel_data_struct_1 = require("../talk/struct/channel-data-struct");
const json_util_1 = require("../util/json-util");
const loco_bson_packet_1 = require("./loco-bson-packet");
class PacketLoginReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(DeviceUUID = '', OAuthToken = '', Appver = '', Os = '', DeviceType = 0, NetType = 0, NetworkMccMnc = '', Language = '', Revision = 0, RevisionData = null, ChannelIdList = [], MaxIdList = [], LastTokenId = bson_1.Long.ZERO, LastChatId = bson_1.Long.ZERO, Lbk = 0, Bg = false) {
        super();
        this.DeviceUUID = DeviceUUID;
        this.OAuthToken = OAuthToken;
        this.Appver = Appver;
        this.Os = Os;
        this.DeviceType = DeviceType;
        this.NetType = NetType;
        this.NetworkMccMnc = NetworkMccMnc;
        this.Language = Language;
        this.Revision = Revision;
        this.RevisionData = RevisionData;
        this.ChannelIdList = ChannelIdList;
        this.MaxIdList = MaxIdList;
        this.LastTokenId = LastTokenId;
        this.LastChatId = LastChatId;
        this.Lbk = Lbk;
        this.Bg = Bg;
    }
    get PacketName() {
        return 'LOGINLIST';
    }
    toBodyJson() {
        let obj = {
            'appVer': this.Appver,
            'prtVer': '1',
            'os': this.Os,
            'lang': this.Language,
            'duuid': this.DeviceUUID,
            'oauthToken': this.OAuthToken,
            'dtype': this.DeviceType,
            'ntype': this.NetType,
            'MCCMNC': this.NetworkMccMnc,
            'revision': this.Revision,
            'rp': null,
            'chatIds': this.ChannelIdList,
            'maxIds': this.MaxIdList,
            'lastTokenId': this.LastTokenId,
            'lbk': this.Lbk,
            'bg': this.Bg
        };
        if (this.LastChatId !== bson_1.Long.ZERO) {
            obj['lastChatId'] = this.LastChatId;
        }
        return obj;
    }
}
exports.PacketLoginReq = PacketLoginReq;
class PacketLoginRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, UserId = bson_1.Long.ZERO, Revision = 0, OpenChatToken = 0, RevisionDetail = '', ChatDataList = []) {
        super(status);
        this.UserId = UserId;
        this.Revision = Revision;
        this.OpenChatToken = OpenChatToken;
        this.RevisionDetail = RevisionDetail;
        this.ChatDataList = ChatDataList;
    }
    get PacketName() {
        return 'LOGINLIST';
    }
    readBodyJson(body) {
        this.UserId = json_util_1.JsonUtil.readLong(body['userId']);
        this.Revision = body['revision'];
        this.RevisionDetail = body['revisionInfo'];
        this.OpenChatToken = body['ltk'];
        this.ChatDataList = [];
        if (body['chatDatas']) {
            let chatDataList = body['chatDatas'];
            for (let rawChatData of chatDataList) {
                this.ChatDataList.push(json_proxy_mapper_1.Serializer.deserialize(rawChatData, channel_data_struct_1.ChannelDataStruct.MAPPER));
            }
        }
    }
}
exports.PacketLoginRes = PacketLoginRes;
