// node-kakao/dist/packet/packet-create-channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketCreateChannelRes = exports.PacketCreateChannelReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const bson_1 = require("bson");
const channel_info_struct_1 = require("../talk/struct/channel-info-struct");
const json_util_1 = require("../util/json-util");
const json_proxy_mapper_1 = require("json-proxy-mapper");
class PacketCreateChannelReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(UserIdList = [], Nickname = '', ProfileURL = '', IsMemoChat = false) {
        super();
        this.UserIdList = UserIdList;
        this.Nickname = Nickname;
        this.ProfileURL = ProfileURL;
        this.IsMemoChat = IsMemoChat;
    }
    get PacketName() {
        return 'CREATE';
    }
    toBodyJson() {
        let obj = {
            'memberIds': this.UserIdList
        };
        if (this.Nickname !== '')
            obj['nickname'] = this.Nickname;
        if (this.ProfileURL !== '')
            obj['profileImageUrl'] = this.ProfileURL;
        if (this.IsMemoChat)
            obj['memoChat'] = this.IsMemoChat;
        return obj;
    }
}
exports.PacketCreateChannelReq = PacketCreateChannelReq;
class PacketCreateChannelRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, ChannelId = bson_1.Long.ZERO, ChatInfo) {
        super(status);
        this.ChannelId = ChannelId;
        this.ChatInfo = ChatInfo;
    }
    get PacketName() {
        return 'CREATE';
    }
    readBodyJson(rawBody) {
        this.ChannelId = json_util_1.JsonUtil.readLong(rawBody['chatId']);
        if (rawBody['chatRoom'])
            this.ChatInfo = json_proxy_mapper_1.Serializer.deserialize(rawBody['chatRoom'], channel_info_struct_1.ChannelInfoStruct.MAPPER);
    }
}
exports.PacketCreateChannelRes = PacketCreateChannelRes;
