// node-kakao/dist/packet/packet-link-deleted.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketLinkDeletedRes = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const chatlog_struct_1 = require("../talk/struct/chatlog-struct");
const bson_1 = require("bson");
const serializer_1 = require("json-proxy-mapper/dist/serializer");
class PacketLinkDeletedRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, LinkId = bson_1.Long.ZERO, OpenToken = 0, Chatlog) {
        super(status);
        this.LinkId = LinkId;
        this.OpenToken = OpenToken;
        this.Chatlog = Chatlog;
    }
    get PacketName() {
        return 'LNKDELETED';
    }
    readBodyJson(body) {
        this.LinkId = body['li'];
        this.OpenToken = body['otk'];
        if (body['chatLog']) {
            this.Chatlog = serializer_1.Serializer.deserialize(body['chatLog'], chatlog_struct_1.ChatlogStruct.MAPPER);
        }
    }
}
exports.PacketLinkDeletedRes = PacketLinkDeletedRes;
