// node-kakao/dist/packet/packet-sync-profile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketSyncProfileRes = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const bson_1 = require("bson");
const open_link_struct_1 = require("../talk/struct/open/open-link-struct");
const json_util_1 = require("../util/json-util");
const json_proxy_mapper_1 = require("json-proxy-mapper");
class PacketSyncProfileRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, ChannelId = bson_1.Long.ZERO, LinkId = bson_1.Long.ZERO, OpenMember) {
        super(status);
        this.ChannelId = ChannelId;
        this.LinkId = LinkId;
        this.OpenMember = OpenMember;
    }
    get PacketName() {
        return 'SYNCLINKPF';
    }
    readBodyJson(rawData) {
        this.ChannelId = json_util_1.JsonUtil.readLong(rawData['c']);
        this.LinkId = json_util_1.JsonUtil.readLong(rawData['li']);
        if (rawData['olu']) {
            this.OpenMember = json_proxy_mapper_1.Serializer.deserialize(rawData['olu'], open_link_struct_1.OpenLinkMemberStruct.MAPPER);
        }
    }
}
exports.PacketSyncProfileRes = PacketSyncProfileRes;
