// node-kakao/dist/packet/packet-react.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketReactRes = exports.PacketReactReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const bson_1 = require("bson");
const open_link_struct_1 = require("../talk/struct/open/open-link-struct");
class PacketReactReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(LinkId = bson_1.Long.ZERO, ReactType = open_link_struct_1.LinkReactionType.NONE) {
        super();
        this.LinkId = LinkId;
        this.ReactType = ReactType;
    }
    get PacketName() {
        return 'REACT';
    }
    toBodyJson() {
        return {
            'li': this.LinkId,
            'rt': this.ReactType
        };
    }
}
exports.PacketReactReq = PacketReactReq;
class PacketReactRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status) {
        super(status);
    }
    get PacketName() {
        return 'REACT';
    }
    readBodyJson(rawData) {
    }
}
exports.PacketReactRes = PacketReactRes;
