// node-kakao/dist/packet/packet-relay-event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketRelayEventRes = exports.PacketRelayEventReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const bson_1 = require("bson");
const chat_type_1 = require("../talk/chat/chat-type");
const relay_event_type_1 = require("../talk/relay/relay-event-type");
class PacketRelayEventReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(LinkId = bson_1.Long.ZERO, ChannelId = bson_1.Long.ZERO, EventType = relay_event_type_1.RelayEventType.UNDEFINED, EventCount = 0, LogId = bson_1.Long.ZERO, Type = chat_type_1.ChatType.Unknown) {
        super();
        this.LinkId = LinkId;
        this.ChannelId = ChannelId;
        this.EventType = EventType;
        this.EventCount = EventCount;
        this.LogId = LogId;
        this.Type = Type;
    }
    get PacketName() {
        return 'RELAYEVENT';
    }
    toBodyJson() {
        return {
            'li': this.LinkId,
            'c': this.ChannelId,
            'et': this.EventType,
            'ec': this.EventCount,
            'logId': this.LogId,
            't': this.Type
        };
    }
}
exports.PacketRelayEventReq = PacketRelayEventReq;
class PacketRelayEventRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, LinkId = bson_1.Long.ZERO, ChannelId = bson_1.Long.ZERO, EventType = relay_event_type_1.RelayEventType.UNDEFINED, EventCount = 0, LogId = bson_1.Long.ZERO, Type = chat_type_1.ChatType.Unknown, AutherId = bson_1.Long.ZERO) {
        super(status);
        this.LinkId = LinkId;
        this.ChannelId = ChannelId;
        this.EventType = EventType;
        this.EventCount = EventCount;
        this.LogId = LogId;
        this.Type = Type;
        this.AutherId = AutherId;
    }
    get PacketName() {
        return 'RELAYEVENT';
    }
    readBodyJson(rawData) {
        this.LinkId = rawData['li'];
        this.ChannelId = rawData['c'];
        this.EventType = rawData['et'];
        this.EventCount = rawData['ec'];
        this.LogId = rawData['logId'];
        this.Type = rawData['t'];
        this.AutherId = rawData['authorId'];
    }
}
exports.PacketRelayEventRes = PacketRelayEventRes;
