// node-kakao/dist/packet/packet-join-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketJoinInfoRes = exports.PacketJoinInfoReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const open_link_struct_1 = require("../talk/struct/open/open-link-struct");
const json_proxy_mapper_1 = require("json-proxy-mapper");
class PacketJoinInfoReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(OpenLinkURL = '', LinkRef = '') {
        super();
        this.OpenLinkURL = OpenLinkURL;
        this.LinkRef = LinkRef;
    }
    get PacketName() {
        return 'JOININFO';
    }
    toBodyJson() {
        return {
            'lu': this.OpenLinkURL,
            'ref': this.LinkRef
        };
    }
}
exports.PacketJoinInfoReq = PacketJoinInfoReq;
class PacketJoinInfoRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, OpenLink) {
        super(status);
        this.OpenLink = OpenLink;
    }
    get PacketName() {
        return 'JOININFO';
    }
    readBodyJson(body) {
        if (body['ol'])
            this.OpenLink = json_proxy_mapper_1.Serializer.deserialize(body['ol'], open_link_struct_1.OpenLinkStruct.MAPPER);
    }
}
exports.PacketJoinInfoRes = PacketJoinInfoRes;
