// node-kakao/dist/packet/packet-join-link.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketJoinLinkRes = exports.PacketJoinLinkReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const __1 = require("..");
const open_link_struct_1 = require("../talk/struct/open/open-link-struct");
const channel_info_struct_1 = require("../talk/struct/channel-info-struct");
const chatlog_struct_1 = require("../talk/struct/chatlog-struct");
const json_proxy_mapper_1 = require("json-proxy-mapper");
class PacketJoinLinkReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(LinkId = __1.Long.ZERO, Referer = '', JoinToken = '', ProfileType = __1.OpenProfileType.MAIN, Nickname = '', ProfilePath = '', ProfileLinkId = __1.Long.ZERO) {
        super();
        this.LinkId = LinkId;
        this.Referer = Referer;
        this.JoinToken = JoinToken;
        this.ProfileType = ProfileType;
        this.Nickname = Nickname;
        this.ProfilePath = ProfilePath;
        this.ProfileLinkId = ProfileLinkId;
    }
    get PacketName() {
        return 'JOINLINK';
    }
    toBodyJson() {
        let obj = {
            'li': this.LinkId,
            'ref': this.Referer,
            'ptp': this.ProfileType
        };
        if (this.JoinToken)
            obj['tk'] = this.JoinToken;
        if (this.ProfileType === __1.OpenProfileType.KAKAO_ANON) {
            if (this.Nickname !== '')
                obj['nn'] = this.Nickname;
            if (this.ProfilePath !== '')
                obj['pp'] = this.ProfilePath;
        }
        else if (this.ProfileType === __1.OpenProfileType.OPEN_PROFILE) {
            obj['pli'] = this.ProfileLinkId;
        }
        return obj;
    }
}
exports.PacketJoinLinkReq = PacketJoinLinkReq;
class PacketJoinLinkRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, LinkInfo, OpenMember, ChatInfo, Chatlog) {
        super(status);
        this.LinkInfo = LinkInfo;
        this.OpenMember = OpenMember;
        this.ChatInfo = ChatInfo;
        this.Chatlog = Chatlog;
    }
    get PacketName() {
        return 'JOINLINK';
    }
    readBodyJson(rawData) {
        if (rawData['ol'])
            this.LinkInfo = json_proxy_mapper_1.Serializer.deserialize(rawData['ol'], open_link_struct_1.OpenLinkStruct.MAPPER);
        if (rawData['olu'])
            this.OpenMember = json_proxy_mapper_1.Serializer.deserialize(rawData['olu'], open_link_struct_1.OpenLinkMemberStruct.MAPPER);
        if (rawData['chatRoom'])
            this.ChatInfo = json_proxy_mapper_1.Serializer.deserialize(rawData['chatRoom'], channel_info_struct_1.ChannelInfoStruct.MAPPER);
        if (rawData['chatLog'])
            this.Chatlog = json_proxy_mapper_1.Serializer.deserialize(rawData['chatLog'], chatlog_struct_1.ChatlogStruct.MAPPER);
    }
}
exports.PacketJoinLinkRes = PacketJoinLinkRes;
