// node-kakao/dist/packet/packet-rewrite.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketRewriteRes = exports.PacketRewriteReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const bson_1 = require("bson");
const feed_type_1 = require("../talk/feed/feed-type");
const chat_type_1 = require("../talk/chat/chat-type");
class PacketRewriteReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(LinkId = bson_1.Long.ZERO, ChannelId = bson_1.Long.ZERO, LogId = bson_1.Long.ZERO, Type = chat_type_1.ChatType.Text, RewriteFeedType = feed_type_1.FeedType.OPENLINK_REWRITE_FEED, ReportChannelLink = '', Category = '') {
        super();
        this.LinkId = LinkId;
        this.ChannelId = ChannelId;
        this.LogId = LogId;
        this.Type = Type;
        this.RewriteFeedType = RewriteFeedType;
        this.ReportChannelLink = ReportChannelLink;
        this.Category = Category;
    }
    get PacketName() {
        return 'REWRITE';
    }
    toBodyJson() {
        let obj = {
            'li': this.LinkId,
            'c': this.ChannelId,
            'logId': this.LogId,
            't': this.Type
        };
        if (this.ReportChannelLink !== '') {
            obj['rcli'] = this.ReportChannelLink;
        }
        if (this.Category !== '') {
            obj['cat'] = this.Category;
        }
        if (this.RewriteFeedType === feed_type_1.FeedType.RICH_CONTENT) {
            obj['ft'] = this.RewriteFeedType;
        }
        return obj;
    }
}
exports.PacketRewriteReq = PacketRewriteReq;
class PacketRewriteRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    get PacketName() {
        return 'REWRITE';
    }
    readBodyJson(body) {
    }
}
exports.PacketRewriteRes = PacketRewriteRes;
