// node-kakao/dist/packet/packet-kick-list-del-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketKickListDelItemRes = exports.PacketKickListDelItemReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const bson_1 = require("bson");
class PacketKickListDelItemReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(LinkId = bson_1.Long.ZERO, ChannelId = bson_1.Long.ZERO, MemberId = bson_1.Long.ZERO) {
        super();
        this.LinkId = LinkId;
        this.ChannelId = ChannelId;
        this.MemberId = MemberId;
    }
    get PacketName() {
        return 'KLDELITEM';
    }
    toBodyJson() {
        return {
            li: this.LinkId,
            c: this.ChannelId,
            kid: this.MemberId
        };
    }
}
exports.PacketKickListDelItemReq = PacketKickListDelItemReq;
class PacketKickListDelItemRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status) {
        super(status);
    }
    get PacketName() {
        return 'KLDELITEM';
    }
    readBodyJson(rawData) {
    }
}
exports.PacketKickListDelItemRes = PacketKickListDelItemRes;
