// node-kakao/dist/packet/packet-channel-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketChannelInfoRes = exports.PacketChannelInfoReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const bson_1 = require("bson");
const channel_info_struct_1 = require("../talk/struct/channel-info-struct");
const json_proxy_mapper_1 = require("json-proxy-mapper");
class PacketChannelInfoReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(ChannelId = bson_1.Long.ZERO) {
        super();
        this.ChannelId = ChannelId;
    }
    get PacketName() {
        return 'CHATINFO';
    }
    toBodyJson() {
        return {
            'chatId': this.ChannelId
        };
    }
}
exports.PacketChannelInfoReq = PacketChannelInfoReq;
class PacketChannelInfoRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, ChatInfo) {
        super(status);
        this.ChatInfo = ChatInfo;
    }
    get PacketName() {
        return 'CHATINFO';
    }
    readBodyJson(rawJson) {
        if (rawJson['chatInfo']) {
            this.ChatInfo = json_proxy_mapper_1.Serializer.deserialize(rawJson['chatInfo'], channel_info_struct_1.ChannelInfoStruct.MAPPER);
        }
    }
}
exports.PacketChannelInfoRes = PacketChannelInfoRes;
