// node-kakao/dist/packet/packet-get-member.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketGetMemberRes = exports.PacketGetMemberReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const __1 = require("..");
const member_struct_1 = require("../talk/struct/member-struct");
const json_proxy_mapper_1 = require("json-proxy-mapper");
const open_link_struct_1 = require("../talk/struct/open/open-link-struct");
class PacketGetMemberReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(ChannelId = __1.Long.ZERO) {
        super();
        this.ChannelId = ChannelId;
    }
    get PacketName() {
        return 'GETMEM';
    }
    toBodyJson() {
        return {
            'chatId': this.ChannelId
        };
    }
}
exports.PacketGetMemberReq = PacketGetMemberReq;
class PacketGetMemberRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, MemberList = []) {
        super(status);
        this.MemberList = MemberList;
    }
    get PacketName() {
        return 'GETMEM';
    }
    readBodyJson(json) {
        this.MemberList = [];
        if (json['members']) {
            let memberList = json['members'];
            for (let rawMem of memberList) {
                if (rawMem[open_link_struct_1.OpenMemberStruct.Mappings.openToken]) {
                    if (rawMem[open_link_struct_1.OpenMemberStruct.Mappings.openToken])
                        this.MemberList.push(json_proxy_mapper_1.Serializer.deserialize(rawMem, open_link_struct_1.OpenMemberStruct.MAPPER));
                }
                else {
                    this.MemberList.push(json_proxy_mapper_1.Serializer.deserialize(rawMem, member_struct_1.MemberStruct.MAPPER));
                }
            }
        }
    }
}
exports.PacketGetMemberRes = PacketGetMemberRes;
