// node-kakao/dist/packet/packet-noti-read.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketMessageNotiReadRes = exports.PacketMessageNotiReadReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const bson_1 = require("bson");
class PacketMessageNotiReadReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(ChannelId = bson_1.Long.ZERO, Watermark = bson_1.Long.ZERO, LinkId) {
        super();
        this.ChannelId = ChannelId;
        this.Watermark = Watermark;
        this.LinkId = LinkId;
    }
    get PacketName() {
        return 'NOTIREAD';
    }
    toBodyJson() {
        let obj = {
            'chatId': this.ChannelId,
            'watermark': this.Watermark
        };
        if (this.LinkId)
            obj['li'] = this.LinkId;
        return obj;
    }
}
exports.PacketMessageNotiReadReq = PacketMessageNotiReadReq;
class PacketMessageNotiReadRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status) {
        super(status);
    }
    get PacketName() {
        return 'NOTIREAD';
    }
    readBodyJson(rawJson) {
    }
}
exports.PacketMessageNotiReadRes = PacketMessageNotiReadRes;
