// node-kakao/dist/packet/packet-info-link.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketInfoLinkRes = exports.PacketInfoLinkReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const open_link_struct_1 = require("../talk/struct/open/open-link-struct");
const json_proxy_mapper_1 = require("json-proxy-mapper");
class PacketInfoLinkReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(LinkIdList = []) {
        super();
        this.LinkIdList = LinkIdList;
    }
    get PacketName() {
        return 'INFOLINK';
    }
    toBodyJson() {
        let obj = {
            'lis': this.LinkIdList
        };
        return obj;
    }
}
exports.PacketInfoLinkReq = PacketInfoLinkReq;
class PacketInfoLinkRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, LinkList = []) {
        super(status);
        this.LinkList = LinkList;
    }
    get PacketName() {
        return 'INFOLINK';
    }
    readBodyJson(rawBody) {
        this.LinkList = [];
        let list = rawBody['ols'];
        if (list) {
            for (let raw of list) {
                this.LinkList.push(json_proxy_mapper_1.Serializer.deserialize(raw, open_link_struct_1.OpenLinkStruct.MAPPER));
            }
        }
    }
}
exports.PacketInfoLinkRes = PacketInfoLinkRes;
