// node-kakao/dist/packet/packet-kick-leave.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketKickLeaveRes = exports.PacketKickLeaveReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const bson_1 = require("bson");
class PacketKickLeaveReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(LinkId = bson_1.Long.ZERO, ChannelId = bson_1.Long.ZERO) {
        super();
        this.LinkId = LinkId;
        this.ChannelId = ChannelId;
    }
    get PacketName() {
        return 'KICKLEAVE';
    }
    toBodyJson() {
        return {
            'li': this.LinkId,
            'c': this.ChannelId
        };
    }
}
exports.PacketKickLeaveReq = PacketKickLeaveReq;
class PacketKickLeaveRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, LastTokenId = bson_1.Long.ZERO) {
        super(status);
        this.LastTokenId = LastTokenId;
    }
    get PacketName() {
        return 'KICKLEAVE';
    }
    readBodyJson(rawBody) {
        this.LastTokenId = rawBody['lastTokenId'];
    }
}
exports.PacketKickLeaveRes = PacketKickLeaveRes;
