// node-kakao/dist/packet/packet-get-trailer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketGetTrailerRes = exports.PacketGetTrailerReq = void 0;
const loco_bson_packet_1 = require("./loco-bson-packet");
const chat_type_1 = require("../talk/chat/chat-type");
class PacketGetTrailerReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(Key = '', Type = chat_type_1.ChatType.Text) {
        super();
        this.Key = Key;
        this.Type = Type;
    }
    get PacketName() {
        return 'GETTRAILER';
    }
    toBodyJson() {
        return {
            'k': this.Key,
            't': this.Type
        };
    }
}
exports.PacketGetTrailerReq = PacketGetTrailerReq;
class PacketGetTrailerRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, Host = '', Port = 0, VHost = '', VHostV6 = '') {
        super(status);
        this.Host = Host;
        this.Port = Port;
        this.VHost = VHost;
        this.VHostV6 = VHostV6;
    }
    get PacketName() {
        return 'GETTRAILER';
    }
    readBodyJson(rawData) {
        this.Host = rawData['h'];
        this.Port = rawData['p'];
        this.VHost = rawData['vh'];
        this.VHostV6 = rawData['vh6'];
    }
}
exports.PacketGetTrailerRes = PacketGetTrailerRes;
