// node-kakao/dist/talk/struct/api/board/channel-post-struct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelPost = exports.BoardEmotionType = exports.PostPermission = exports.PostType = void 0;
var PostType;
(function (PostType) {
    PostType["TEXT"] = "TEXT";
    PostType["POLL"] = "POLL";
    PostType["FILE"] = "FILE";
    PostType["IMAGE"] = "IMAGE";
    PostType["VIDEO"] = "VIDEO";
    PostType["SCHEDULE"] = "SCHEDULE";
})(PostType = exports.PostType || (exports.PostType = {}));
var PostPermission;
(function (PostPermission) {
})(PostPermission = exports.PostPermission || (exports.PostPermission = {}));
var BoardEmotionType;
(function (BoardEmotionType) {
    BoardEmotionType["LIKE"] = "LIKE";
})(BoardEmotionType = exports.BoardEmotionType || (exports.BoardEmotionType = {}));
var ChannelPost;
(function (ChannelPost) {
    let ContentType;
    (function (ContentType) {
        ContentType["TEXT"] = "text";
        ContentType["MENTION"] = "user";
        ContentType["EVERYONE_MENTION"] = "user_all";
    })(ContentType = ChannelPost.ContentType || (ChannelPost.ContentType = {}));
    let PollItemType;
    (function (PollItemType) {
        PollItemType["TEXT"] = "text";
    })(PollItemType = ChannelPost.PollItemType || (ChannelPost.PollItemType = {}));
})(ChannelPost = exports.ChannelPost || (exports.ChannelPost = {}));
