// node-kakao/dist/store/store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsyncIdInstanceStore = exports.IdInstanceStore = exports.AsyncIdStore = exports.IdStore = void 0;
class IdStore {
    constructor() {
        this.map = new Map();
    }
    has(key) {
        return this.map.has(this.convertKey(key));
    }
    get(key) {
        return this.map.get(this.convertKey(key)) || null;
    }
    set(key, value) {
        return this.map.set(this.convertKey(key), value);
    }
    delete(key) {
        return this.map.delete(this.convertKey(key));
    }
    clear() {
        return this.map.clear();
    }
    values() {
        return this.map.values();
    }
    convertKey(key) {
        return key.toString();
    }
}
exports.IdStore = IdStore;
class AsyncIdStore {
    constructor() {
        this.map = new Map();
    }
    has(key) {
        return this.map.has(this.convertKey(key));
    }
    async get(key) {
        return this.map.get(this.convertKey(key)) || null;
    }
    getFromMap(key) {
        return this.map.get(this.convertKey(key));
    }
    set(key, value) {
        return this.map.set(this.convertKey(key), value);
    }
    delete(key) {
        return this.map.delete(this.convertKey(key));
    }
    clear() {
        return this.map.clear();
    }
    values() {
        return this.map.values();
    }
    convertKey(key) {
        return key.toString();
    }
}
exports.AsyncIdStore = AsyncIdStore;
class IdInstanceStore extends IdStore {
    get(key) {
        let val = super.get(key);
        if (!val) {
            val = this.createInstanceFor(key);
            if (!val)
                return null;
            this.set(key, val);
        }
        return val;
    }
}
exports.IdInstanceStore = IdInstanceStore;
class AsyncIdInstanceStore extends AsyncIdStore {
    async get(key) {
        let val = await super.get(key);
        if (!val) {
            val = await this.createInstanceFor(key);
            if (!val)
                return null;
            this.set(key, val);
        }
        return val;
    }
}
exports.AsyncIdInstanceStore = AsyncIdInstanceStore;
