// node-kakao/dist/api/api-header-decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHeaderDecorator = exports.AHeaderDecorator = exports.BasicHeaderDecorator = void 0;
const client_config_provider_1 = require("../config/client-config-provider");
class BasicHeaderDecorator {
    constructor(configProvider) {
        this.configProvider = configProvider;
    }
    get UserAgent() {
        return `KT/${this.configProvider.Configuration.version} Wd/${this.configProvider.Configuration.osVersion} ${this.configProvider.Configuration.language}`;
    }
    fillHeader(header) {
        header['Accept'] = '*/*';
        header['Accept-Language'] = this.configProvider.Configuration.language;
        header['User-Agent'] = this.UserAgent;
    }
}
exports.BasicHeaderDecorator = BasicHeaderDecorator;
BasicHeaderDecorator.INSTANCE = new BasicHeaderDecorator(new client_config_provider_1.DefaultClientConfigProvider());
class AHeaderDecorator {
    constructor(configProvider) {
        this.configProvider = configProvider;
    }
    fillHeader(header) {
        header['A'] = `${this.configProvider.Configuration.agent}/${this.configProvider.Configuration.version}/${this.configProvider.Configuration.language}`;
    }
}
exports.AHeaderDecorator = AHeaderDecorator;
AHeaderDecorator.INSTANCE = new AHeaderDecorator(new client_config_provider_1.DefaultClientConfigProvider());
class CHeaderDecorator {
    constructor() {
    }
    fillHeader(header) {
    }
}
exports.CHeaderDecorator = CHeaderDecorator;
