// node-kakao/dist/api/open-upload-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenUploadApi = void 0;
const web_api_client_1 = require("./web-api-client");
class OpenUploadApi extends web_api_client_1.WebApiClient {
    get Scheme() {
        return 'https';
    }
    get Host() {
        return 'up-api1-kage.kakao.com';
    }
    async uploadLinkImage(name, file, contentType = 'image/jpeg') {
        return this.requestMultipart('POST', 'up/open-chat-profile/', { 'file_1': { value: file, options: { filename: name, contentType: contentType } } });
    }
    async uploadLinkPostImage(name, file, contentType = 'image/jpeg') {
        return this.requestMultipart('POST', 'up/open-chat-profile-post/', { 'file_1': { value: file, options: { filename: name, contentType: contentType } } });
    }
    getOriginalLinkImageURL(accessKey) {
        return `http://open.kakaocdn.net/dn/${accessKey}/img.jpg`;
    }
    getSmallLinkImageURL(accessKey) {
        return `http://open.kakaocdn.net/dn/${accessKey}/img_s.jpg`;
    }
    getLargeLinkImageURL(accessKey) {
        return `http://open.kakaocdn.net/dn/${accessKey}/img_l.jpg`;
    }
}
exports.OpenUploadApi = OpenUploadApi;
