// node-kakao/dist/config/client-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultConfiguration = void 0;
exports.DefaultConfiguration = {
    locoBookingURL: 'booking-loco.kakao.com',
    locoBookingPort: 443,
    locoPEMPublicKey: `-----BEGIN PUBLIC KEY-----\nMIIBIDANBgkqhkiG9w0BAQEFAAOCAQ0AMIIBCAKCAQEApElgRBx+g7sniYFW7LE8ivrwXShKTRFV8lXNItMXbN5QSC8vJ/cTSOTS619Xv5Zx7xXJIk4EKxtWesEGbgZpEUP2xQ+IeH9oz0JxayEMvvD1nVNAWgpWE4pociEoArsK7qY3YwXb1CiDHo9hojLv7djbo3cwXvlyMh4TUrX2RjCZPlVJxk/LVjzcl9ohJLkl3eoSrf0AE4kQ9mk3+raEhq5Dv+IDxKYX+fIytUWKmrQJusjtre9oVUX5sBOYZ0dzez/XapusEhUWImmB6mciVXfRXQ8IK4IH6vfNyxMSOTfLEhRYN2SMLzplAYFiMV536tLS3VmG5GJRdkpDubqPeQIBAw==\n-----END PUBLIC KEY-----`,
    agent: 'win32',
    version: '3.1.4',
    appVersion: '3.1.4.2500',
    osVersion: '10.0',
    deviceType: 2,
    netType: 0,
    mccmnc: '999',
    countryIso: 'KR',
    language: 'ko',
    subDevice: true,
    deviceModel: '',
    xvcSeedList: ['HEATH', 'DEMIAN'],
    loginTokenSeedList: ['PITT', 'INORAN']
};
