// node-kakao/dist/talk/struct/channel-meta-struct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelClientMetaType = exports.ChannelMetaType = void 0;
var ChannelMetaType;
(function (ChannelMetaType) {
    ChannelMetaType[ChannelMetaType["UNDEFINED"] = 0] = "UNDEFINED";
    ChannelMetaType[ChannelMetaType["NOTICE"] = 1] = "NOTICE";
    ChannelMetaType[ChannelMetaType["GROUP"] = 2] = "GROUP";
    ChannelMetaType[ChannelMetaType["TITLE"] = 3] = "TITLE";
    ChannelMetaType[ChannelMetaType["PROFILE"] = 4] = "PROFILE";
    ChannelMetaType[ChannelMetaType["TV"] = 5] = "TV";
    ChannelMetaType[ChannelMetaType["PRIVILEGE"] = 6] = "PRIVILEGE";
    ChannelMetaType[ChannelMetaType["TV_LIVE"] = 7] = "TV_LIVE";
    ChannelMetaType[ChannelMetaType["PLUS_BACKGROUND"] = 8] = "PLUS_BACKGROUND";
    ChannelMetaType[ChannelMetaType["LIVE_TALK_INFO"] = 11] = "LIVE_TALK_INFO";
    ChannelMetaType[ChannelMetaType["LIVE_TALK_COUNT"] = 12] = "LIVE_TALK_COUNT";
    ChannelMetaType[ChannelMetaType["OPEN_CHANNEL_CHAT"] = 13] = "OPEN_CHANNEL_CHAT";
    ChannelMetaType[ChannelMetaType["BOT"] = 14] = "BOT";
})(ChannelMetaType = exports.ChannelMetaType || (exports.ChannelMetaType = {}));
var ChannelClientMetaType;
(function (ChannelClientMetaType) {
    ChannelClientMetaType["UNDEFINED"] = "undefined";
    ChannelClientMetaType["NAME"] = "name";
    ChannelClientMetaType["IMAGE_PATH"] = "image_path";
    ChannelClientMetaType["FAVORITE"] = "favorite";
    ChannelClientMetaType["PUSH_SOUND"] = "push_sound";
    ChannelClientMetaType["CHAT_HIDE"] = "chat_hide";
    ChannelClientMetaType["FULL_IMAGE_URL"] = "full_image_url";
    ChannelClientMetaType["IMAGE_URL"] = "imageUrl";
})(ChannelClientMetaType = exports.ChannelClientMetaType || (exports.ChannelClientMetaType = {}));
