// node-kakao/dist/talk/struct/channel-info-struct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelInfoStruct = void 0;
const chatlog_struct_1 = require("./chatlog-struct");
const member_struct_1 = require("./member-struct");
const json_proxy_mapper_1 = require("json-proxy-mapper");
var ChannelInfoStruct;
(function (ChannelInfoStruct) {
    ChannelInfoStruct.Mappings = {
        channelId: 'chatId',
        type: 'type',
        activeMemberCount: 'activeMembersCount',
        newMessageCount: 'newMessageCount',
        lastUpdatedAt: 'lastUpdatedAt',
        lastMessage: 'lastMessage',
        lastLogId: 'lastLogId',
        lastSeenLogId: 'lastSeenLogId',
        lastChatLog: 'lastChatLog',
        metadata: 'meta',
        displayMemberList: 'displayMembers',
        pushAlert: 'pushAlert',
        channelMetaList: 'chatMetas',
        isDirectChat: 'directChat',
        linkId: 'li',
        openToken: 'otk'
    };
    ChannelInfoStruct.ConvertMap = {
        displayMemberList: new json_proxy_mapper_1.Converter.Array(member_struct_1.DisplayMemberStruct.Mappings),
        lastChatLog: new json_proxy_mapper_1.Converter.Object(chatlog_struct_1.ChatlogStruct.Mappings)
    };
    ChannelInfoStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(ChannelInfoStruct.Mappings, ChannelInfoStruct.ConvertMap);
})(ChannelInfoStruct = exports.ChannelInfoStruct || (exports.ChannelInfoStruct = {}));
