// node-kakao/dist/talk/struct/channel-board-meta-struct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelBoardMetaType = exports.ChannelBoardMetaStruct = void 0;
const json_proxy_mapper_1 = require("json-proxy-mapper");
var ChannelBoardMetaStruct;
(function (ChannelBoardMetaStruct) {
    ChannelBoardMetaStruct.Mappings = {
        'type': 't',
        'content': 'ct',
        'boardRevision': 'br',
        'userRevision': 'ur'
    };
    ChannelBoardMetaStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(ChannelBoardMetaStruct.Mappings);
})(ChannelBoardMetaStruct = exports.ChannelBoardMetaStruct || (exports.ChannelBoardMetaStruct = {}));
var ChannelBoardMetaType;
(function (ChannelBoardMetaType) {
    ChannelBoardMetaType[ChannelBoardMetaType["NONE"] = 0] = "NONE";
    ChannelBoardMetaType[ChannelBoardMetaType["FLOATING_NOTICE"] = 1] = "FLOATING_NOTICE";
    ChannelBoardMetaType[ChannelBoardMetaType["SIDE_NOTICE"] = 2] = "SIDE_NOTICE";
    ChannelBoardMetaType[ChannelBoardMetaType["BADGE"] = 3] = "BADGE";
})(ChannelBoardMetaType = exports.ChannelBoardMetaType || (exports.ChannelBoardMetaType = {}));
