// node-kakao/dist/talk/struct/member-struct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemberStruct = exports.DisplayMemberStruct = void 0;
const json_proxy_mapper_1 = require("json-proxy-mapper");
var DisplayMemberStruct;
(function (DisplayMemberStruct) {
    DisplayMemberStruct.Mappings = {
        userId: 'userId',
        nickname: 'nickName',
        profileImageUrl: 'pi'
    };
    DisplayMemberStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(DisplayMemberStruct.Mappings);
})(DisplayMemberStruct = exports.DisplayMemberStruct || (exports.DisplayMemberStruct = {}));
var MemberStruct;
(function (MemberStruct) {
    MemberStruct.Mappings = {
        userId: 'userId',
        nickname: 'nickName',
        profileImageUrl: 'profileImageUrl',
        fullProfileImageUrl: 'fullProfileImageUrl',
        originalProfileImageUrl: 'originalProfileImageUrl',
        type: 'type',
        accountId: 'accountId',
        linkedService: 'linkedService',
        statusMessage: 'statusMessage',
        openToken: 'opt',
        memberType: 'lmt',
        profileType: 'ptp',
        linkId: 'li',
        pv: 'pv'
    };
    MemberStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(MemberStruct.Mappings);
})(MemberStruct = exports.MemberStruct || (exports.MemberStruct = {}));
