// node-kakao/dist/talk/struct/auth/auth-api-struct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthStatusCode = void 0;
var AuthStatusCode;
(function (AuthStatusCode) {
    AuthStatusCode[AuthStatusCode["LOGIN_FAILED_REASON"] = 12] = "LOGIN_FAILED_REASON";
    AuthStatusCode[AuthStatusCode["LOGIN_FAILED"] = 30] = "LOGIN_FAILED";
    AuthStatusCode[AuthStatusCode["MOBILE_UNREGISTERED"] = 32] = "MOBILE_UNREGISTERED";
    AuthStatusCode[AuthStatusCode["DEVICE_NOT_REGISTERED"] = -100] = "DEVICE_NOT_REGISTERED";
    AuthStatusCode[AuthStatusCode["ANOTHER_LOGON"] = -101] = "ANOTHER_LOGON";
    AuthStatusCode[AuthStatusCode["DEVICE_REGISTER_FAILED"] = -102] = "DEVICE_REGISTER_FAILED";
    AuthStatusCode[AuthStatusCode["INVALID_DEVICE_REGISTER"] = -110] = "INVALID_DEVICE_REGISTER";
    AuthStatusCode[AuthStatusCode["INCORRECT_PASSCODE"] = -111] = "INCORRECT_PASSCODE";
    AuthStatusCode[AuthStatusCode["PASSCODE_REQUEST_FAILED"] = -112] = "PASSCODE_REQUEST_FAILED";
    AuthStatusCode[AuthStatusCode["ACCOUNT_RESTRICTED"] = -997] = "ACCOUNT_RESTRICTED";
})(AuthStatusCode = exports.AuthStatusCode || (exports.AuthStatusCode = {}));
