// node-kakao/dist/talk/struct/auth/login-access-data-struct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoginAccessDataStruct = void 0;
const json_util_1 = require("../../../util/json-util");
const json_proxy_mapper_1 = require("json-proxy-mapper");
var LoginAccessDataStruct;
(function (LoginAccessDataStruct) {
    let mappings = {
        status: 'status',
        message: 'message',
        userId: 'userId',
        countryISO: 'countryIso',
        countryCode: 'countryCode',
        accountId: 'accountId',
        logonServerTime: 'server_time',
        resetUserData: 'resetUserData',
        accessToken: 'access_token',
        refreshToken: 'refresh_token',
        tokenType: 'token_type',
        autoLoginEmail: 'autoLoginAccountId',
        displayAccountId: 'displayAccountId',
        mainDevice: 'mainDeviceAgentName',
        mainDeviceAppVersion: 'mainDeviceAppVersion'
    };
    let convertMap = {
        userId: json_util_1.JsonUtil.LongConverter
    };
    LoginAccessDataStruct.MAPPER = new json_proxy_mapper_1.ObjectMapper(mappings, convertMap);
})(LoginAccessDataStruct = exports.LoginAccessDataStruct || (exports.LoginAccessDataStruct = {}));
