// node-kakao/dist/talk/open/open-link-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenChannelType = exports.OpenProfileType = exports.OpenMemberType = exports.OpenLinkType = void 0;
var OpenLinkType;
(function (OpenLinkType) {
    OpenLinkType[OpenLinkType["PROFILE"] = 1] = "PROFILE";
    OpenLinkType[OpenLinkType["CHANNEL"] = 2] = "CHANNEL";
})(OpenLinkType = exports.OpenLinkType || (exports.OpenLinkType = {}));
var OpenMemberType;
(function (OpenMemberType) {
    OpenMemberType[OpenMemberType["OWNER"] = 1] = "OWNER";
    OpenMemberType[OpenMemberType["NONE"] = 2] = "NONE";
    OpenMemberType[OpenMemberType["MANAGER"] = 4] = "MANAGER";
    OpenMemberType[OpenMemberType["BOT"] = 8] = "BOT";
})(OpenMemberType = exports.OpenMemberType || (exports.OpenMemberType = {}));
var OpenProfileType;
(function (OpenProfileType) {
    OpenProfileType[OpenProfileType["MAIN"] = 1] = "MAIN";
    OpenProfileType[OpenProfileType["KAKAO_ANON"] = 2] = "KAKAO_ANON";
    OpenProfileType[OpenProfileType["KAKAO_ANON_2"] = 4] = "KAKAO_ANON_2";
    OpenProfileType[OpenProfileType["UNKNOWN_1"] = 8] = "UNKNOWN_1";
    OpenProfileType[OpenProfileType["OPEN_PROFILE"] = 16] = "OPEN_PROFILE";
})(OpenProfileType = exports.OpenProfileType || (exports.OpenProfileType = {}));
var OpenChannelType;
(function (OpenChannelType) {
    OpenChannelType[OpenChannelType["UNKNOWN"] = 0] = "UNKNOWN";
    OpenChannelType[OpenChannelType["DIRECT"] = 1] = "DIRECT";
    OpenChannelType[OpenChannelType["GROUP"] = 2] = "GROUP";
})(OpenChannelType = exports.OpenChannelType || (exports.OpenChannelType = {}));
