// node-kakao/dist/packet/checkin/packet-check-in.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketCheckInRes = exports.PacketCheckInReq = void 0;
const loco_bson_packet_1 = require("../loco-bson-packet");
const bson_1 = require("bson");
class PacketCheckInReq extends loco_bson_packet_1.LocoBsonRequestPacket {
    constructor(UserId = bson_1.Long.ZERO, Os = '', NetType = 0, Appver = '', NetworkMccMnc = '', language = '', CountryIso = '', UseSub = true) {
        super();
        this.UserId = UserId;
        this.Os = Os;
        this.NetType = NetType;
        this.Appver = Appver;
        this.NetworkMccMnc = NetworkMccMnc;
        this.language = language;
        this.CountryIso = CountryIso;
        this.UseSub = UseSub;
    }
    get PacketName() {
        return 'CHECKIN';
    }
    toBodyJson() {
        let obj = {
            userId: this.UserId,
            os: this.Os,
            ntype: this.NetType,
            appVer: this.Appver,
            MCCMNC: this.NetworkMccMnc,
            lang: this.language,
            useSub: this.UseSub
        };
        if (this.CountryIso)
            obj['countryISO'] = this.CountryIso;
        return obj;
    }
}
exports.PacketCheckInReq = PacketCheckInReq;
class PacketCheckInRes extends loco_bson_packet_1.LocoBsonResponsePacket {
    constructor(status, Host = '', Port = 0, CacheExpire = -1) {
        super(status);
        this.Host = Host;
        this.Port = Port;
        this.CacheExpire = CacheExpire;
    }
    get PacketName() {
        return 'CHECKIN';
    }
    readBodyJson(body) {
        this.Host = body['host'];
        this.Port = body['port'];
        this.CacheExpire = body['cacheExpire'];
    }
}
exports.PacketCheckInRes = PacketCheckInRes;
